garde/src/i18n.rs

```diff
@@ -0,0 +1,250 @@
+    fn phone_number_invalid(&self) -> String;
+
+    /// Rule: `phone_number`
+    fn phone_number_invalid_with_error(&self, error: &str) -> String;
+
+    /// Rule: `ip`
+    fn ip_invalid(&self, kind: &str) -> String;
+
+    /// Rule: `matches`
+    fn matches_field_mismatch(&self, field: &str) -> String;
+
+    /// Rule: `email`
+    fn email_invalid(&self, error: &str) -> String;
+
+    /// Rule: `ascii`
+    fn ascii_invalid(&self) -> String;
+
+    /// Rule: `alphanumeric`
+    fn alphanumeric_invalid(&self) -> String;
+
+    /// Rule: `required`
+    fn required_not_set(&self) -> String;
+}
+
+macro_rules! i18n {
+    ($handler:ident $(, $($args:expr),*)?) => {
+        $crate::i18n::I18N.with_borrow(|i18n| {
+            let i18n = i18n.as_deref().unwrap_or(&$crate::i18n::DefaultI18n);
+            i18n.$handler($($($args),*)?)
+        })
+    };
+}
+
+/// Default implementation of [`I18n`] which provides english error messages.
+pub struct DefaultI18n;
+
+impl I18n for DefaultI18n {
+    fn length_lower_than(&self, min: usize) -> String {
+        format!("length is lower than {min}")
+    }
+
+    fn length_greater_than(&self, max: usize) -> String {
+        format!("length is greater than {max}")
+    }
+
+    fn range_lower_than(&self, min: &str) -> String {
+        format!("lower than {min}")
+    }
+
+    fn range_greater_than(&self, max: &str) -> String {
+        format!("greater than {max}")
+    }
+
+    fn credit_card_invalid(&self, error: &str) -> String {
+        format!("not a valid credit card number: {error}")
+    }
+
+    fn pattern_no_match(&self, pattern: &str) -> String {
+        format!("does not match pattern /{pattern}/")
+    }
+
+    fn contains_missing(&self, pattern: &str) -> String {
+        format!("does not contain \"{pattern}\"")
+    }
+
+    fn url_invalid(&self, error: &str) -> String {
+        format!("not a valid url: {error}")
+    }
+
+    fn prefix_missing(&self, pattern: &str) -> String {
+        format!("value does not begin with \"{pattern}\"")
+    }
+
+    fn suffix_missing(&self, pattern: &str) -> String {
+        format!("does not end with \"{pattern}\"")
+    }
+
+    fn phone_number_invalid(&self) -> String {
+        "not a valid phone number".to_owned()
+    }
+
+    fn phone_number_invalid_with_error(&self, error: &str) -> String {
+        format!("not a valid phone number: {error}")
+    }
+
+    fn ip_invalid(&self, kind: &str) -> String {
+        format!("not a valid {kind} address")
+    }
+
+    fn matches_field_mismatch(&self, field: &str) -> String {
+        format!("does not match {field} field")
+    }
+
+    fn email_invalid(&self, error: &str) -> String {
+        format!("not a valid email: {error}")
+    }
+
+    fn ascii_invalid(&self) -> String {
+        "not ascii".to_owned()
+    }
+
+    fn alphanumeric_invalid(&self) -> String {
+        "not alphanumeric".to_owned()
+    }
+
+    fn required_not_set(&self) -> String {
+        "not set".to_owned()
+    }
```

Comment:
Same here, The user can see that the field is unset, why not tell them that it is required, to make it more clear WHY we want them to change that.
```suggestion
    fn required_not_set(&self, field_name: &str) -> String {
        format!("{field_name} is required.")
    }
```